src/MessageHandler/ProcessFileHandler.php

```diff
@@ -0,0 +1,166 @@
+        $this->logger = $datasetFileHasherLogger;
+        $this->datastore = $datastore;
+        $this->messageBus = $messageBus;
+    }
+
+    /**
+     * Destructor for the handler to always flush.
+     */
+    public function __destruct()
+    {
+        $this->entityManager->flush();
+        foreach ($this->messages as $message) {
+            $this->messageBus->dispatch($message);
+        }
+    }
+
+    /**
+     * Invoke function to process a file.
+     *
+     * @param ProcessFile $processFile The HasFile message to be handled.
+     */
+    public function __invoke(ProcessFile $processFile)
+    {
+        $fileId = $processFile->getFileId();
+        $file = $this->fileRepository->find($fileId);
+
+        if (!$file instanceof File) {
+            $this->logger->alert(sprintf('File with ID: %d was not found!', $fileId));
+            return;
+        }
+
+        $fileset = $file->getFileset();
+        $datasetSubmission = $fileset->getDatasetSubmission();
+        $dataset = $datasetSubmission->getDataset();
+        $udi = $dataset->getUdi();
+        $loggingContext = array(
+            'fileId' => $fileId,
+            'dataset_id' => $dataset->getId(),
+            'udi' => $udi,
+            'dataset_submission_id' => $datasetSubmission->getId(),
+        );
+
+        $file->setStatus(File::FILE_IN_PROGRESS);
+        $filePath = $file->getPhysicalFilePath();
+        $fileStream = fopen($filePath, 'r');
+        try {
+            $fileHash = StreamInfo::calculateHash(array('fileStream' => $fileStream));
+            $file->setFileSha256Hash($fileHash);
+        } catch (\Exception $e) {
+            $this->logger->warning(sprintf('Unable to hash file. Message: %s', $e->getMessage()));
+        }
+
+        try {
+            $newFileDestination = $this->datastore->addFile(
+                ['fileStream' => $fileStream],
+                str_replace(':', '.', $udi) . DIRECTORY_SEPARATOR . $file->getFilePathName()
+            );
+            $file->setPhysicalFilePath($newFileDestination);
+        } catch (\Exception $exception) {
+            $this->logger->error(sprintf('Unable to add file to datastore. Message: "%s"', $exception->getMessage()), $loggingContext);
+            $file->setStatus(File::FILE_ERROR);
+            return;
+        }
+
+        // File virus Scan
+        $this->messages[] = new ScanFileForVirus($fileId, $loggingContext['udi']);
+        $this->logger->info("Enqueuing virus scan for file: {$file->getFilePathName()}.", $loggingContext);
+
+        $file->setStatus(File::FILE_DONE);
+
+        if ($fileset->isDone()) {
+            $datasetSubmissionId = $datasetSubmission->getId();
+            $fileIds = array();
+            foreach ($fileset->getProcessedFiles() as $file) {
+                $fileIds[] = $file->getId();
+            }
+            // Dispatch message to zip files
+            $zipFiles = new ZipDatasetFiles($fileIds, $datasetSubmissionId);
```

Comment:
We are building the zip after processing every file?

Reply:
Ok, tested this again. I didn't see the $fileset->isDone() condition above, this is fine 👍